Sources/Crypto/BCrypt/BCryptDigest.swift

```diff
@@ -1,37 +1,118 @@
+import libbcrypt
+import Random
+
 /// Creates and verifies BCrypt hashes. Normally you will not need to initialize one of these classes and you will
 /// use the global `BCrypt` convenience instead.
 ///
 ///     try BCrypt.hash("vapor", cost: 4)
 ///
 /// See `BCrypt` for more information.
 public final class BCryptDigest {
+
     /// Creates a new `BCryptDigest`. Use the global `BCrypt` convenience variable.
     public init() { }
 
+    private enum Algorithm: String, RawRepresentable {
+        /// older version
+        case _2a = "$2a$"
+        /// format specific to the crypt_blowfish BCrypt implementation, identical to `2b` in all but name.
+        case _2y = "$2y$"
+        /// latest revision of the official BCrypt algorithm, current default
+        case _2b = "$2b$"
+
+        /// Revision's length, including the `$` symbols
+        var revisionCount: Int {
+            return 4
+        }
+
+        /// Salt's length (includes revision and cost info)
+        var fullSaltCount: Int {
+            return 29
+        }
+
+        /// Checksum's length
+        var checksumCount: Int {
+            return 31
+        }
+
+        /// Salt's length (does NOT include neither revision nor cost info)
+        static var saltCount: Int {
+            return 22
+        }
+    }
+
     /// Creates a BCrypt digest for the supplied data.
     ///
-    ///     try BCrypt.hash("vapor", cost: 4)
-    ///
-    ///  If a `salt` is not supplied, a random one will be generated. You can use a custom salt if desired.
+    /// Salt must be provided
     ///
     ///     try BCrypt.hash("vapor", cost: 12, salt: "passwordpassword")
     ///
     /// - parameters:
-    ///     - plaintext: Plaintext data to digest.
+    ///     - plaintextData: Plaintext data to digest.
     ///     - cost: Desired complexity. Larger `cost` values take longer to hash and verify.
     ///     - salt: Optional salt for this hash. If omitted, a random salt will be generated.
-    ///             The salt must be 16-bytes.
+    ///             The salt must be 16-bytes if provided by the user (without cost, revision data)
     /// - throws: `CryptoError` if hashing fails or if data conversion fails.
     /// - returns: BCrypt hash for the supplied plaintext data.
-    public func hash(_ plaintext: LosslessDataConvertible, cost: Int = 12, salt: LosslessDataConvertible? = nil) throws -> Data {
-        let config = try BCryptConfig(
-            version: .two(.y),
-            cost: cost,
-            salt: salt?.convertToData() ?? CryptoRandom().generateData(count: 16)
+    public func hash(_ plaintextData: LosslessDataConvertible, cost: UInt = 12, salt saltData: LosslessDataConvertible) throws -> String {
```

Comment:
One tiny thing i just noticed here: `Int` -> `UInt`. We should probably keep this as `Int` just to keep the API convenient (using UInt can make things pretty annoying sometimes). A lot of other places in Vapor that only support unsigned integers tend to just use `Int` to make peoples' lives easier. But I could go either way on this. If we do change to `UInt`, we should make sure no other places in Crypto are accepting `Int` where it must be unsigned to keep things consistent and predictable.

Reply:
I was going to ask about it :P